### PR TITLE
Feature/message input on before send

### PR DESCRIPTION
**Pull request by Chesteer89**
Follow-up ticket for docs should be created (code sample)

### CODE REVIEW

**Review by mateusz-cichosz — APPROVED** [review]

File: chat-components/database/src/main/java/com/pubnub/components/DefaultDatabase.kt [hunk 1]
```diff
@@ -37,7 +36,6 @@ import com.pubnub.components.data.message.action.DefaultMessageActionDao
     ChannelCustomDataConverter::class,
     DateConverter::class,
     MemberDataCustomConverter::class,
-    CustomContentConverter::class,
```
Comment (Chesteer89) [inline chat-components/database/src/main/java/com/pubnub/components/DefaultDatabase.kt:40]: @kazydek `Converter` removed, it's used directly in `DBMessage` class

File: chat-components/database/src/main/java/com/pubnub/components/data/message/CustomContentConverter.kt [hunk 2]
```diff
@@ -2,18 +2,26 @@ package com.pubnub.components.data.message
...
 import androidx.room.TypeConverter
 import com.google.gson.GsonBuilder
-import com.google.gson.reflect.TypeToken
 
 class CustomContentConverter {
-    private val gson = GsonBuilder()
-        .enableComplexMapKeySerialization()
-        .create()
 
     @TypeConverter
```
Comment (Chesteer89) [inline chat-components/database/src/main/java/com/pubnub/components/data/message/CustomContentConverter.kt:8]: @kazydek Converter updated, we can prepare a sample code with the usage

File: chat-components/database/src/main/java/com/pubnub/components/data/message/DBMessage.kt [hunk 3]
```diff
@@ -22,15 +19,15 @@ data class DBMessage(
     override val id: String,
     override val text: String,
     override val contentType: String? = null,
-    override val content: DBCustomContent? = null,
+    @field:TypeConverters(CustomContentConverter::class)
```
Comment (Chesteer89) [inline chat-components/database/src/main/java/com/pubnub/components/data/message/DBMessage.kt:22]: @kazydek `custom` and `content` fields marked with annotation (to parse json to `Any` type)

File: chat-components/src/main/java/com/pubnub/components/chat/network/data/NetworkHistoryMessage.kt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package com.pubnub.components.chat.network.data
+
+import com.pubnub.api.models.consumer.history.PNFetchMessageItem
+
+typealias NetworkHistoryMessage = PNFetchMessageItem
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/network/data/NetworkHistoryMessage.kt:5]: @kazydek Additional `typealias` added to keep it independent in code

File: chat-components/src/main/java/com/pubnub/components/chat/network/mapper/NetworkMessageActionHistoryMapper.kt [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
 
-import com.pubnub.api.models.consumer.history.PNFetchMessageItem
+import com.pubnub.components.chat.network.data.NetworkHistoryMessage
 import com.pubnub.components.data.message.action.DBMessageAction
 import com.pubnub.framework.data.ChannelId
 import com.pubnub.framework.mapper.MapperWithId
 
-class NetworkMessageActionHistoryMapper : MapperWithId<PNFetchMessageItem, Array<DBMessageAction>> {
-    override fun map(id: ChannelId, input: PNFetchMessageItem): Array<DBMessageAction> {
+class NetworkMessageActionHistoryMapper : MapperWithId<NetworkHistoryMessage, Array<DBMessageAction>> {
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/network/mapper/NetworkMessageActionHistoryMapper.kt:8]: @kazydek type updated

File: chat-components/src/main/java/com/pubnub/components/chat/network/mapper/NetworkMessageHistoryMapper.kt [hunk 6]
```diff
@@ -2,32 +2,31 @@ package com.pubnub.components.chat.network.mapper
...
 import com.pubnub.components.chat.network.data.NetworkMessagePayload
-import com.pubnub.components.data.message.DBCustomContent
 import com.pubnub.components.data.message.DBMessage
 import com.pubnub.framework.data.ChannelId
 import com.pubnub.framework.mapper.MapperWithId
 import com.pubnub.framework.util.toJson
 
 class NetworkMessageHistoryMapper(private val mapper: MapperManager) :
-    MapperWithId<PNFetchMessageItem, DBMessage> {
+    MapperWithId<NetworkHistoryMessage, DBMessage> {
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/network/mapper/NetworkMessageHistoryMapper.kt:13]: @kazydek type updated

File: chat-components/src/main/java/com/pubnub/components/chat/network/mapper/NetworkMessageMapper.kt [hunk 7]
```diff
@@ -18,9 +17,9 @@ class NetworkMessageMapper(private val mapper: MapperManager) : Mapper<NetworkMe
             id = messagePayload.id,
             text = messagePayload.text,
             contentType = messagePayload.contentType,
-            content = messagePayload.content as DBCustomContent?,
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/network/mapper/NetworkMessageMapper.kt:21]: @kazydek cast to `DBCustomContent` removed. Now if the developer wants to get his own object he needs to call `message.content.mapToObject(MyCustomType::class)`

File: chat-components/src/main/java/com/pubnub/components/chat/network/paging/MessageRemoteMediator.kt [hunk 8]
```diff
@@ -15,7 +16,7 @@ import java.io.IOException
 @OptIn(ExperimentalPagingApi::class, DelicateCoroutinesApi::class)
 class MessageRemoteMediator constructor(
     private val channelId: ChannelId,
-    private val service: MessageService<DBMessage>,
+    private val service: MessageService<NetworkMessagePayload>,
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/network/paging/MessageRemoteMediator.kt:19]: @kazydek type updated

File: chat-components/src/main/java/com/pubnub/components/chat/provider/ChatProvider.kt [hunk 9]
```diff
@@ -158,6 +157,7 @@ fun WithServices(
     CompositionLocalProvider(
         LocalMessageService provides DefaultMessageService(
             LocalPubNub.current,
+            LocalUser.current,
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/provider/ChatProvider.kt:160]: @kazydek `LocalUser` parameter passed to `DefaultMessageService`

File: chat-components/src/main/java/com/pubnub/components/chat/service/message/DefaultMessageService.kt [hunk 10]
```diff
@@ -39,7 +42,7 @@ class DefaultMessageService(
     private val logger: Logger,
     private val coroutineScope: CoroutineScope = GlobalScope,
     private val dispatcher: CoroutineDispatcher = Dispatchers.IO,
-) : MessageService<DBMessage> {
+) : MessageService<NetworkMessagePayload> {
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/service/message/DefaultMessageService.kt:45]: @kazydek Type updated

File: chat-components/src/main/java/com/pubnub/components/chat/service/message/DefaultMessageService.kt [hunk 11]
```diff
@@ -64,42 +67,47 @@ class DefaultMessageService(
     /**
      * Send a message to all subscribers of a channel.
      *
-     * @param id ID of channel to send a message to
-     * @param message object to send
+     * @param channelId ID of channel to send a message to
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/service/message/DefaultMessageService.kt:70]: @kazydek Parameter name updated, jdocs updated

File: chat-components/src/main/java/com/pubnub/components/chat/service/message/MessageService.kt [hunk 12]
```diff
@@ -12,7 +12,7 @@ interface MessageService<Data : Message> {
 
     @Suppress("NAME_SHADOWING")
     suspend fun send(
-        id: ChannelId,
+        channelId: ChannelId,
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/service/message/MessageService.kt:15]: @kazydek Parameter name updated

File: chat-components/src/main/java/com/pubnub/components/chat/service/message/MessageService.kt [hunk 13]
```diff
@@ -31,7 +31,7 @@ interface MessageService<Data : Message> {
 }
 
 val LocalMessageService =
-    staticCompositionLocalOf<MessageService<DBMessage>> { throw MessageServiceNotInitializedException() }
+    staticCompositionLocalOf<MessageService<NetworkMessagePayload>> { throw MessageServiceNotInitializedException() }
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/service/message/MessageService.kt:34]: @kazydek Type updated

File: chat-components/src/main/java/com/pubnub/components/chat/ui/component/input/MessageInputExt.kt [hunk 14]
```diff
@@ -28,6 +29,7 @@ fun MessageInput(
     typingIndicatorContent: @Composable ColumnScope.(List<TypingUi>) -> Unit = { typing ->
         TypingIndicatorContent(typing)
     },
+    onBeforeSend: ((String) -> NetworkMessagePayload)? = null,
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/ui/component/input/MessageInputExt.kt:32]: @kazydek `onBeforeSend` method added

File: chat-components/src/main/java/com/pubnub/components/chat/ui/mapper/message/DBMessageMapper.kt [hunk 15]
```diff
@@ -25,6 +25,7 @@ class DBMessageMapper(
             reactions = input.actions.toUi(),
             contentType = input.message.contentType ?: "default",
             content = input.message.content,
+            custom = input.custom,
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/ui/mapper/message/DBMessageMapper.kt:28]: @kazydek `custom` parsed

File: chat-components/src/main/java/com/pubnub/components/chat/ui/mapper/message/DomainMessageMapper.kt [hunk 16]
```diff
@@ -14,9 +14,9 @@ class DomainMessageMapper : Mapper<MessageUi.Data, DBMessageWithActions> {
                 id = input.uuid,
                 text = input.text,
                 contentType = input.contentType,
-                content = null,
+                content = input.content,
```
Comment (Chesteer89) [inline chat-components/src/main/java/com/pubnub/components/chat/ui/mapper/message/DomainMessageMapper.kt:17]: @kazydek `custom` and `content` mapped

File: chat-components/src/test/java/com/pubnub/components/MessageRemoteMediatorUnitTest.kt [hunk 17]
```diff
@@ -38,7 +39,7 @@ class MessageRemoteMediatorUnitTest {
     private val testCoroutineScope = TestScope(testCoroutineDispatcher)
 
     @MockK(relaxed = true)
-    lateinit var service: MessageService<DBMessage>
+    lateinit var service: MessageService<NetworkMessagePayload>
```
Comment (Chesteer89) [inline chat-components/src/test/java/com/pubnub/components/MessageRemoteMediatorUnitTest.kt:42]: @kazydek type updated

File: chat-components/ui/src/main/java/com/pubnub/components/chat/ui/component/message/MessageUi.kt [hunk 18]
```diff
@@ -18,7 +18,8 @@ sealed class MessageUi {
         val isDelivered: Boolean,
         val reactions: List<ReactionUi>,
         val contentType: String = "default",
-        val content: Map<String, Any?>? = null,
+        val content: Any? = null,
```
Comment (Chesteer89) [inline chat-components/ui/src/main/java/com/pubnub/components/chat/ui/component/message/MessageUi.kt:21]: @kazydek `custom` field added to `MessageUi`

File: chat-components/database/src/main/java/com/pubnub/components/data/message/CustomContentConverter.kt (hunk 2 above)
Comment (kazydek) [reply chat-components/database/src/main/java/com/pubnub/components/data/message/CustomContentConverter.kt:8]: yup, as part of another task

**Pull request merged**